File: server/ai-helper/skills/ppt-master/scripts/svg_text_wrap.py
#!/usr/bin/env python3
"""Best-effort SVG text wrapping guard for ppt-master outputs.

This script is intentionally conservative: it only rewrites plain, long
single-line <text> nodes into <tspan> lines when the estimated text width would
overflow the local card/column. It avoids foreignObject so the SVG remains PPT
export friendly.
"""

from __future__ import annotations

import argparse
import math
import re
from pathlib import Path
from xml.etree import ElementTree as ET


SVG_NS = "http://www.w3.org/2000/svg"
ET.register_namespace("", SVG_NS)


def tag_name(element: ET.Element) -> str:
    return element.tag.rsplit("}", 1)[-1]


def first_float(value: str | None, default: float = 0.0) -> float:
    if not value:
        return default
    match = re.search(r"-?\d+(?:\.\d+)?", value)
    return float(match.group(0)) if match else default


def parse_viewbox(root: ET.Element) -> tuple[float, float]:
    raw = root.get("viewBox") or ""
    parts = [float(p) for p in raw.split() if re.match(r"^-?\d+(?:\.\d+)?$", p)]
    if len(parts) == 4:
        return max(parts[2], 1), max(parts[3], 1)
    return first_float(root.get("width"), 1600), first_float(root.get("height"), 900)


def visual_weight(char: str) -> float:
    if char.isspace():
        return 0.32
    if "\u4e00" <= char <= "\u9fff" or "\u3000" <= char <= "\u303f":
        return 1.02
    if char in "，。；：、（）【】《》“”‘’":
        return 0.8
    if char.isupper() or char.isdigit():
        return 0.62
    return 0.54


def estimate_width(text: str, font_size: float) -> float:
    return sum(visual_weight(ch) for ch in text) * font_size


def split_tokens(text: str) -> list[str]:
    tokens: list[str] = []
    buffer = ""
    for ch in text:
        if ch.isspace():
            if buffer:
                tokens.append(buffer)
                buffer = ""
            tokens.append(ch)
        elif "\u4e00" <= ch <= "\u9fff" or "\u3000" <= ch <= "\u303f":
            if buffer:
                tokens.append(buffer)
                buffer = ""
            tokens.append(ch)
        else:
            buffer += ch
    if buffer:
        tokens.append(buffer)
    return tokens


def wrap_text(text: str, font_size: float, max_width: float) -> list[str]:
    tokens = split_tokens(text.strip())
    lines: list[str] = []
    current = ""
    for token in tokens:
        candidate = (current + token).strip() if token.isspace() else current + token
        if current and estimate_width(candidate, font_size) > max_width:
            lines.append(current.strip())
            current = token.strip()
        else:
            current = candidate
    if current.strip():
        lines.append(current.strip())
    return lines


def local_rect_width(text_el: ET.Element, siblings: list[ET.Element], x: float, y: float) -> float | None:
    best: tuple[float, float] | None = None
    for sibling in siblings:
        if tag_name(sibling) != "rect":
            continue
        rx = first_float(sibling.get("x"))
        ry = first_float(sibling.get("y"))
        rw = first_float(sibling.get("width"))
        rh = first_float(sibling.get("height"))
        if rw <= 0 or rh <= 0:
            continue
        if rx <= x <= rx + rw and ry <= y <= ry + rh:
            area = rw * rh
            available = rx + rw - x - max(18, first_float(text_el.get("font-size"), 18) * 0.8)
            if available > 0 and (best is None or area < best[0]):
                best = (area, available)
    return best[1] if best else None


def inferred_max_width(
    text_el: ET.Element,
    siblings: list[ET.Element],
    canvas_w: float,
    canvas_h: float,
) -> float:
    x = first_float(text_el.get("x"))
    y = first_float(text_el.get("y"))
    font_size = first_float(text_el.get("font-size"), 18)
    rect_width = local_rect_width(text_el, siblings, x, y)
    if rect_width and rect_width > font_size * 7:
        return rect_width

    margin = max(48, canvas_w * 0.045)
    if x < canvas_w * 0.48:
        return max(font_size * 8, canvas_w * 0.48 - x - margin * 0.35)
    if x < canvas_w * 0.72:
        return max(font_size * 8, canvas_w * 0.72 - x - margin * 0.35)
    return max(font_size * 8, canvas_w - x - margin)


def should_wrap(text_el: ET.Element, max_width: float) -> bool:
    if list(text_el):
        return False
    if (text_el.get("text-anchor") or "").lower() in {"middle", "end"}:
        return False
    text = (text_el.text or "").strip()
    if len(text) < 18:
        return False
    font_size = first_float(text_el.get("font-size"), 18)
    return estimate_width(text, font_size) > max_width


def wrap_file(svg_path: Path, *, dry_run: bool = False) -> int:
    tree = ET.parse(svg_path)
    root = tree.getroot()
    canvas_w, canvas_h = parse_viewbox(root)
    parent_map = {child: parent for parent in root.iter() for child in list(parent)}
    changed = 0

    for text_el in list(root.iter()):
        if tag_name(text_el) != "text":
            continue
        parent = parent_map.get(text_el)
        siblings = list(parent) if parent is not None else []
        font_size = first_float(text_el.get("font-size"), 18)
        max_width = inferred_max_width(text_el, siblings, canvas_w, canvas_h)
        if not should_wrap(text_el, max_width):
            continue
        lines = wrap_text(text_el.text or "", font_size, max_width)
        if len(lines) <= 1:
            continue
        text_el.text = None
        base_x = text_el.get("x") or "0"
        line_gap = math.ceil(font_size * 1.22)
        for idx, line in enumerate(lines):
            tspan = ET.Element(f"{{{SVG_NS}}}tspan")
            tspan.set("x", base_x)
            if idx:
                tspan.set("dy", str(line_gap))
            else:
                tspan.set("dy", "0")
            tspan.text = line
            text_el.append(tspan)
        changed += 1

    if changed and not dry_run:
        tree.write(svg_path, encoding="unicode", xml_declaration=False)
    return changed


def collect_svg_files(path: Path) -> list[Path]:
    if path.is_file() and path.suffix.lower() == ".svg":
        return [path]
    svg_dir = path / "svg_output" if (path / "svg_output").exists() else path
    return sorted(svg_dir.glob("*.svg"))


def main() -> int:
    parser = argparse.ArgumentParser(description="Wrap long single-line SVG text into PPT-safe tspans.")
    parser.add_argument("path", type=Path, help="SVG file, svg_output directory, or ppt-master project path.")
    parser.add_argument("--dry-run", action="store_true")
    args = parser.parse_args()

    files = collect_svg_files(args.path)
    if not files:
        print(f"No SVG files found under {args.path}")
        return 1
    total = 0
    for svg in files:
        count = wrap_file(svg, dry_run=args.dry_run)
        total += count
        if count:
            print(f"{svg}: wrapped {count} text node(s)")
    print(f"SVG text wrap complete: {total} text node(s) wrapped across {len(files)} file(s)")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: server/ai-helper/skills/ppt-master/scripts/svg_text_wrap.ts
#!/usr/bin/env tsx
import fs from 'fs';
import path from 'path';

function attr(attrs: string, name: string): string | undefined {
  const match = attrs.match(new RegExp(`\\b${name}\\s*=\\s*(['"])(.*?)\\1`, 'i'));
  return match ? match[2] : undefined;
}
function firstNumber(value: string | undefined, fallback = 0): number {
  const match = String(value || '').match(/-?\d+(?:\.\d+)?/);
  return match ? Number(match[0]) : fallback;
}
function decode(value: string): string {
  return value.replace(/&lt;/g, '<').replace(/&gt;/g, '>').replace(/&quot;/g, '"').replace(/&apos;/g, "'").replace(/&amp;/g, '&');
}
function escapeText(value: string): string {
  return value.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;');
}
function weight(ch: string): number {
  if (/\s/.test(ch)) return 0.32;
  if (/[\u4e00-\u9fff\u3000-\u303f]/.test(ch)) return 1.02;
  if ('，。；：、（）【】《》“”‘’'.includes(ch)) return 0.8;
  if (/[A-Z0-9]/.test(ch)) return 0.62;
  return 0.54;
}
function estimate(text: string, fontSize: number): number { return Array.from(text).reduce((s, c) => s + weight(c), 0) * fontSize; }
function tokens(text: string): string[] {
  const out: string[] = []; let buf = '';
  for (const ch of Array.from(text)) {
    if (/\s/.test(ch) || /[\u4e00-\u9fff\u3000-\u303f]/.test(ch)) { if (buf) out.push(buf); buf = ''; out.push(ch); }
    else buf += ch;
  }
  if (buf) out.push(buf);
  return out;
}
function lines(text: string, fontSize: number, maxWidth: number): string[] {
  const out: string[] = []; let cur = '';
  for (const token of tokens(text.trim())) {
    const candidate = /\s/.test(token) ? `${cur}${token}`.trim() : `${cur}${token}`;
    if (cur && estimate(candidate, fontSize) > maxWidth) { out.push(cur.trim()); cur = token.trim(); }
    else cur = candidate;
  }
  if (cur.trim()) out.push(cur.trim());
  return out;
}
function canvasWidth(content: string): number {
  const root = content.match(/<svg\b([^>]*)>/i)?.[1] || '';
  const parts = String(attr(root, 'viewBox') || '').split(/[\s,]+/).map(Number).filter(Number.isFinite);
  return parts.length === 4 ? Math.max(parts[2], 1) : firstNumber(attr(root, 'width'), 1280);
}
function maxWidth(attrs: string, width: number): number {
  const x = firstNumber(attr(attrs, 'x'), 0);
  const fs = firstNumber(attr(attrs, 'font-size'), 18);
  const margin = Math.max(48, width * 0.045);
  if (x < width * 0.48) return Math.max(fs * 8, width * 0.48 - x - margin * 0.35);
  if (x < width * 0.72) return Math.max(fs * 8, width * 0.72 - x - margin * 0.35);
  return Math.max(fs * 8, width - x - margin);
}
function wrapContent(content: string): { content: string; changed: number } {
  const width = canvasWidth(content); let changed = 0;
  const next = content.replace(/<text\b([^>]*)>([^<]+)<\/text>/gi, (raw, attrs: string, body: string) => {
    const anchor = (attr(attrs, 'text-anchor') || '').toLowerCase();
    if (anchor === 'middle' || anchor === 'end') return raw;
    const text = decode(body).trim();
    if (text.length < 18) return raw;
    const fs = firstNumber(attr(attrs, 'font-size'), 18);
    const mw = maxWidth(attrs, width);
    if (estimate(text, fs) <= mw) return raw;
    const split = lines(text, fs, mw);
    if (split.length <= 1) return raw;
    const x = attr(attrs, 'x') || '0'; const gap = Math.ceil(fs * 1.22);
    changed += 1;
    return `<text${attrs}>${split.map((line, i) => `<tspan x="${x}" dy="${i ? gap : 0}">${escapeText(line)}</tspan>`).join('')}</text>`;
  });
  return { content: next, changed };
}
function main(): void {
  const projectDir = path.resolve(process.argv[2] || '');
  if (!process.argv[2]) throw new Error('Usage: svg_text_wrap.ts <project_path>');
  const svgDir = path.join(projectDir, 'svg_output');
  const files = fs.readdirSync(svgDir).filter((x) => x.toLowerCase().endsWith('.svg')).sort();
  let changed = 0;
  for (const file of files) {
    const abs = path.join(svgDir, file);
    const original = fs.readFileSync(abs, 'utf8');
    const wrapped = wrapContent(original);
    if (wrapped.changed && wrapped.content !== original) { fs.writeFileSync(abs, wrapped.content, 'utf8'); changed += wrapped.changed; }
  }
  console.log(JSON.stringify({ ok: true, files: files.length, changed }));
}
try { main(); } catch (err) { console.error(err instanceof Error ? err.message : String(err)); process.exit(1); }
